ocfweb/docs/docs/staff/backend/kubernetes.md

```diff
@@ -0,0 +1,204 @@
+            *VIP*          HAProxy          |                    Ingress         Pod        |
+        ------------      ----------  ✘ SSL / Host: templates   ---------     -----------   |
+REQ --> |keepalived| -->  |deadlock|    -->   -------------->   |Worker1| --> |Templates|   |
+        ------------      ----------        \                   ---------     -----------   |
+                                            |                                               |
+                           HAProxy          |                    Ingress         Pod        |
+                          ----------        |                   ---------     -----------   |
+                          |  coup  |        |                   |Worker3|     | grafana |   |
+                          ----------        |                   ---------     -----------   |
+                                            -------------------------------------------------
+```
+
+All three Kubernetes masters are running an instance of [HAProxy][haproxy].
+Furthermore, the masters are all running `keepalived`. The traffic for any
+Kubernetes HTTP service will go through the current `keepalived` master, which
+holds the virtual IP for all Kubernetes services. The `keepalived` master is
+randomly chosen but will move hosts in the case of failure.  `HAProxy` will
+terminate ssl and pass the request on to a worker running [Ingress
+Nginx][ingress-nginx].  Right now ingress is running as a [NodePort][nodeport]
+service on all workers (Note: we can easily change this to be a subset of
+workers if our cluster scales such that this is no longer feasible).  The
+ingress worker will inspect the `Host` header and forward the request on to the
+appropriate pod where the request is finally processed. _Do note that the
+target pod is not necessarily on the same worker that routed the traffic_.
+
+
+### Why didn't we use MetalLB?
+
+`MetalLB` was created so a bare-metal Kubernetes cluster could use `Type:
+LoadBalancer` in Service definitions. The problem is, in `L2` mode, it takes a
+pool of IPs and puts your service on a random IP in that pool. How one makes
+DNS work in this configuration is completely unspecified. We would need to
+dynamically update our DNS, which to me sounds like a myriad of outages waiting
+to happen. `L3` mode would require the OCF dedicating a router to Kubernetes.
+
+
+### Why don't we copy Marathan and specify one port per service?
+
+In our current Marathon configuration, we give each service a port on the load
+balancer and traffic coming into that port is routed accordingly. First, in
+Kubernetes we would emulate this behavior using `NodePort` services, and all
+Kubernetes documentation discourages this. Second, it's ugly. Every time we add
+a new service we need to modify the load balancer configuration in Puppet. With
+our Kubernetes configuration we can add unlimited HTTP services without
+touching Puppet.
```

Comment:
Where does HAProxy end up getting its config from instead?

Reply:
HAProxy is configured in puppet. All it needs to know is where workers running ingress live (right now that is all nodes)

Reply:
Ingress workers route based on Host headers, which is why our current config only scales for HTTP services. If we wanted to expose TCP services, we would probably need to modify HAProxy